WIP: FEATURE: New reference editor (DD Sprint)

mhsdesign opened the pull request:
WIP


| Screenshots from the Dresden sprint | |
|--------|--------|
| ![reference-editor-multiple-delete](https://github.com/user-attachments/assets/51eb31e9-e254-4065-869c-36121ee2365c) | ![reference-editor-releation-properties](https://github.com/user-attachments/assets/e9257b23-491f-4651-bfaa-9aa1bcc8688e) |
| ![reference-editor-select-new](https://github.com/user-attachments/assets/12b226b3-1d41-49a0-be74-8ed64ad21aa6) | ![reference-editor-selected-in-inspector](https://github.com/user-attachments/assets/6c7ac6dd-7006-4a98-848c-009f3808d495) | 


Stand 21.10.2025

## Ziel
- References besser auswählen können (in Secondary Editor als Tree, mit Suche und Nodefilter)
- Referenceproperties bearbeitbar machen

## Yaml
- references eigene Ebene
- Jede Reference hat einen Namen und kann properties haben

## Inspector

### Ui
- Anzeige aller References
- Oberhalb der References sind ausgegraut Buttons: 
	- Hinzufügen (+): Öffnet Secondary Editor mit Treeview
	- Springen zu (Pfeil): Springt zur Source der ausgewählten Referenz
	- Properties editieren (Irgendeine Darstellung einer Krake/Corona/...): Öffnet das Modal mit Referenceproperty Editor
	- Löschen (Papierkorb): Löscht ausgewählte Referenz
- Verschieben der References möglich durch Drag&Drop
- Auswahl einer Reference durch Klick möglich
	- In dem Fall werden Buttons oberhalb der References aktiv
- Haben die References im Schema keine properties hinterlegt, wird der Edit-Button ausgeblendet
- Sind in einer Reference properties ausgefüllt, zeige ein Icon (Lsite) auf der Reference an
	- Kann später evtl. abgelöst werden durch die Read-Only-Vorschau des RefererencePropertyEditors (siehe Nice to have)
- (Nice to have) Bei Hover:
	- werden in der Reference die Buttons zum Editieren, Löschen, .... angezeigt (um klar zu machen, dass diese bei Select aktiv werden)
	- Es wird eine Read-Only Vorschau des RefererencePropertyEditors angezeigt, um zu zeigen, welche Properties es gibt und z.B. bei Images eine Vorschau zu ermöglichen

### Backend
- API-Schnittstelle liefert Liste aller References zurück
- Jede Reference enthält:
	- breadcrums
	- uri
	- sourceNodeId
	- label
	- icon
	- properties
	- Schema (wenn vorhanden)

## Reference selektieren

### Ui 
- Öffnet sich im Secondary Editor
- Wenn geöffnet, wird im Inspector Hinzufügen, Editieren und Springen zu ReferenceNode deaktiviert
- Anzeige:
	- Tree aller verfügbaren Nodes
	- Bereits ausgewählte References sind hervorgehoben und haben neben sich ein Löschen-Icon
	- Optional (erst mal weglassen): Es gibt ein Properties-Editieren-Icon neben den References, mit dem sich der ReferencePropertyEditor öffnet
- Filtern nach Nodetypes mgl.
- Suche mgl.
- Klick auf Node:
	- Node wird den References hinzugefügt, sollte sie noch nicht daz gehören
	- Man soll NICHT via Klick auf eine Node die entsprechende Reference wieder entfernen, wenn sie schon ausgewählt ist

### Backend
- Alles schon von LinkEditor als Controller verfügbar --> kopieren
- Erst bei Apply werden alle hinzugefügten References dann auch tatsächlich in der DB gesetzt

## Reference Properties editieren

### Ui
- Modal
- Links wird sourceNode angezeigt
- In der Mitte properties der Reference in Editor
- Rechts wird reference angezeigt, hier kann man auch durch alle References blättern

mhsdesign committed:
TASK: Introduce initial test for get-references-summary

mhsdesign committed:
TASK: Introduce initial test for get-references-summary

mhsdesign committed:
WIP: Initial copy of `GetReferencesSummary`

mhsdesign committed:
WIP: Initial copy of `GetReferencesSummary`

sjuschkat committed:
WIP: 'GetReferencesSummary' Test adjusted, API created to fetch references in ReferencesEditor

sjuschkat committed:
WIP: 'GetReferencesSummary' Test adjusted, API created to fetch references in ReferencesEditor

mhsdesign committed:
WIP: Add policy

mhsdesign committed:
WIP: Add policy

mhsdesign committed:
WIP: Add policy lol 2

mhsdesign committed:
WIP: Add policy lol 2

so-grimm committed:
WIP: add behat test scenario for multiple references, fetch reference properties

so-grimm committed:
WIP: add behat test scenario for multiple references, fetch reference properties

Commit:
WIP: Display references from the new API EP

73nici committed:
WIP: Display references from the new API EP

Commit:
WIP: More style adjustments for the Reference Item

73nici committed:
WIP: More style adjustments for the Reference Item

sjuschkat committed:
WIP: 'GetReferencesSummary' Test adjusted, properties changed for 'GetReferencesSummaryQueryResult', schema, constraints and properties of references fetched in 'GetReferencesSummaryQueryHandler'

sjuschkat committed:
WIP: 'GetReferencesSummary' Test adjusted, properties changed for 'GetReferencesSummaryQueryResult', schema, constraints and properties of references fetched in 'GetReferencesSummaryQueryHandler'

73nici committed:
WIP: Add hover actions

Commit:
WIP: Add hover actions

so-grimm committed:
WIP: implement endpoint for GetChildrenForTreeNode and added tests

so-grimm committed:
WIP: implement endpoint for GetChildrenForTreeNode and added tests

so-grimm committed:
WIP: implement endpoint for GetTree and added tests for getTree including constraints

so-grimm committed:
WIP: implement endpoint for GetTree and added tests for getTree including constraints

73nici committed:
WIP: Add neos-ui-reference-editor-custom-node-tree

Commit:
WIP: Add neos-ui-reference-editor-custom-node-tree

Commit:
WIP: Use props.identifier for request

73nici committed:
WIP: Use props.identifier for request

mhsdesign committed:
WIP: Inline `ReferencesList`

mhsdesign committed:
WIP: Inline `ReferencesList`

mhsdesign committed:
WIP: Introduce global reference editor state handling for opening relation properties dialog

mhsdesign committed:
WIP: Introduce global reference editor state handling for opening relation properties dialog

Commit:
WIP: Fix empty referenceIds

73nici committed:
WIP: Fix empty referenceIds

sjuschkat committed:
WIP: Class Reference created to wrap single reference in GetReferencesSummaryQuery, pulled constraints and propertySchema up into GetReferencesSummaryQuery, referenceIds as queryParameter removed, test adjusted

sjuschkat committed:
WIP: Class Reference created to wrap single reference in GetReferencesSummaryQuery, pulled constraints and propertySchema up into GetReferencesSummaryQuery, referenceIds as queryParameter removed, test adjusted

mhsdesign committed:
TASK: Allow to que additional changes (future reference and subtree change) via `commitChange` when using the inspector

mhsdesign committed:
TASK: Allow to que additional changes (future reference and subtree change) via `commitChange` when using the inspector

mhsdesign committed:
WIP: Show reference property editors

mhsdesign committed:
WIP: Show reference property editors

mhsdesign committed:
WIP: Que changes from the reference editor

mhsdesign committed:
WIP: Que changes from the reference editor

mhsdesign committed:
WIP: Wip handle inspector discard event to flush references editor -> implement proper references listing with delete button

mhsdesign committed:
WIP: Wip handle inspector discard event to flush references editor -> implement proper references listing with delete button

mhsdesign committed:
WIP: Wip add new dedicated Reference Change type

mhsdesign committed:
WIP: Wip add new dedicated Reference Change type

laurahaenel committed:
WIP: secondary editor

laurahaenel committed:
WIP: secondary editor

laurahaenel committed:
TASK: show selected nodes in secondary editor

laurahaenel committed:
TASK: show selected nodes in secondary editor

mhsdesign committed:
WIP: reference property editing

mhsdesign committed:
WIP: reference property editing

laurahaenel committed:
WIP: structure secondary editor

laurahaenel committed:
WIP: structure secondary editor

laurahaenel committed:
WIP: fix install

laurahaenel committed:
WIP: fix install

mhsdesign committed:
WIP: deduplicate qued reference changes (only one is to be committed)

mhsdesign committed:
WIP: deduplicate qued reference changes (only one is to be committed)

mhsdesign committed:
WIP: persist reference properties

mhsdesign committed:
WIP: persist reference properties

mhsdesign committed:
WIP: add initial editor test

mhsdesign committed:
WIP: add initial editor test

laurahaenel committed:
WIP: subscribe to selected nodes & select node action

laurahaenel committed:
WIP: subscribe to selected nodes & select node action

mhsdesign committed:
WIP: add prevent closing when reference property editing is dirty

mhsdesign committed:
WIP: add prevent closing when reference property editing is dirty

laurahaenel committed:
TASK: mock api

laurahaenel committed:
TASK: mock api

mhsdesign committed:
WIP: fix laura

mhsdesign committed:
WIP: fix laura

mhsdesign committed:
WIP: fix apply disabled state and fix dialog zindex

mhsdesign committed:
WIP: fix apply disabled state and fix dialog zindex

mhsdesign committed:
WIP: add initial references count from server

this will be used to prevent content shifting

mhsdesign committed:
WIP: add initial references count from server

this will be used to prevent content shifting

mhsdesign committed:
WIP: bugfix editor never turns transient values to initial State which leads to a later discard to reset the presentation to the initial server values

mhsdesign committed:
WIP: bugfix editor never turns transient values to initial State which leads to a later discard to reset the presentation to the initial server values

laurahaenel committed:
WIP: loading state on ref values

laurahaenel committed:
WIP: loading state on ref values

laurahaenel committed:
WIP: fix state handling on transient and initial values for pending presentation values

laurahaenel committed:
WIP: fix state handling on transient and initial values for pending presentation values

laurahaenel committed:
WIP: add test for discard and apply pending changes

laurahaenel committed:
WIP: add test for discard and apply pending changes

mhsdesign committed:
WIP: copy GetNodeSummary Query from link editor for reference editor

this will be used to calculate what the presentation of a new selected reference will look like

mhsdesign committed:
WIP: copy GetNodeSummary Query from link editor for reference editor

this will be used to calculate what the presentation of a new selected reference will look like

mhsdesign committed:
WIP: use dedicated icon for reference property editing

mhsdesign committed:
WIP: use dedicated icon for reference property editing

mhsdesign committed:
WIP: try out simple `IconCard` as presentation (todo move to shared package)

mhsdesign committed:
WIP: try out simple `IconCard` as presentation (todo move to shared package)

mhsdesign committed:
TASK: Update references editor to also use `Flow\Route` and `AbstractQueryController` like the new `LinkEditor`

adjustments for 28cb6ef3648f059ab47ad2abf18757bdc8e96f93

mhsdesign committed:
TASK: Update references editor to also use `Flow\Route` and `AbstractQueryController` like the new `LinkEditor`

adjustments for 28cb6ef3648f059ab47ad2abf18757bdc8e96f93

mhsdesign commented:
TODO 3.6.2026

```
- [ ] Required Features from previous editor
  - [ ] Search over tree in secondary editor
  - [ ] Grey-out disallowed nodes according to configured type in node tree (the nodes must not be hidden as there might be valid selectable child nodes in the sub tree)
  - [ ] Implement sorting of references in primary inspector
- [ ] Hover navigation with solid background and 3rd button to navigate to node
- [ ] Move node selection into dialog and add preview list with actions (drag, delete, edit) as right column into dialog as well
- [ ] Add visual indications that properties exist on the reference edge (as composed icon ontop of the node type icon -> see existing Neos Ui component)
- [ ] Prevent accidental deletion / unselection of reference edge as this could have properties -> only deletable via preview list where reference property indicator is available.

- [ ] JS Application Design
  - [ ] Reference editor must work in secondary editor
  - [ ] Reference inspector editor state lost when switching tabs during editing
```

EDIT: Our current progress and thread resides now here https://neos-project.slack.com/archives/C0U0KEGDQ/p1780475996782389 there is also an interactive todo list.

laurahaenel commented:
Drag n Drop use either of that maybe? Decide later, I just leave it here
1. https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API
2. https://dndkit.com/react/quickstart/

so-grimm committed:
WIP: switch dimensionSpacePoint from legacy to jsonString

- replace `DimensionSpacePoint::fromLegacyDimensionArray` with `DimensionSpacePoint::fromJsonString` for tree node getters in ReferenceEditor and LinkEditor
- check that dimensionSpacePoint is set in all tree node getters for LinkEditor and ReferenceEditor
- `GetNodeSummaryController` - replace LinkEditor use statements with ReferenceEditor
- adjust behat tests for ReferenceEditor and LinkEditor to use jsonString format for dimensionSpacePoint
- rename `dimensionValues` to `dimensionSpacePoint`
- deprecate `DimensionValues` and `DimensionCombination` in `packages/neos-ts-interfaces/src/index.ts`
- add `ContentDimensionValue` and `DimensionSpacePoint` and implement converter function `createDimensionSpacePointFromLegacyDimension()`
- convert old `dimensionValues` to new `DimensionSpacePoint` in ts files

so-grimm committed:
WIP: fix tree loading for multiple selected nodes at deeper levels

- foreach loops for finding siblings were overriding `hasUnloadedChildren`, setting it to `true``even if the children were already fully loaded
- added check for children in `setHasUnloadedChildren()` to ensure it doesn't get reset for already loaded children
